File: android/src/main/java/expo/modules/superwallexpo/json/CustomerInfo.kt
package expo.modules.superwallexpo.json

import com.superwall.sdk.models.customer.CustomerInfo
import com.superwall.sdk.models.customer.NonSubscriptionTransaction
import com.superwall.sdk.models.customer.SubscriptionTransaction
import com.superwall.sdk.store.abstractions.product.receipt.LatestPeriodType
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Date

// Mirrors the @SerialName values on `LatestPeriodType`. Exhaustive `when` so a new
// SDK case is a compile error rather than a silently-incorrect string.
private fun LatestPeriodType.toJsonString(): String =
  when (this) {
    LatestPeriodType.TRIAL -> "trial"
    LatestPeriodType.CODE -> "code"
    LatestPeriodType.SUBSCRIPTION -> "subscription"
    LatestPeriodType.PROMOTIONAL -> "promotional"
    LatestPeriodType.WINBACK -> "winback"
    LatestPeriodType.REVOKED -> "revoked"
  }

private val dateFormatter: DateTimeFormatter = DateTimeFormatter.ISO_INSTANT

private fun Date.toIsoString(): String =
  this.toInstant().atZone(ZoneId.systemDefault()).format(dateFormatter)

fun CustomerInfo.toJson(): Map<String, Any?> {
  val map = mutableMapOf<String, Any?>()
  map["userId"] = this.userId
  map["subscriptions"] = this.subscriptions.map { it.toJson() }
  map["nonSubscriptions"] = this.nonSubscriptions.map { it.toJson() }
  map["entitlements"] = this.entitlements.map { it.toJson() }
  return map
}

fun SubscriptionTransaction.toJson(): Map<String, Any?> {
  val map = mutableMapOf<String, Any?>()
  map["transactionId"] = this.transactionId
  map["productId"] = this.productId
  map["purchaseDate"] = this.purchaseDate.toIsoString()
  map["willRenew"] = this.willRenew
  map["isRevoked"] = this.isRevoked
  map["isInGracePeriod"] = this.isInGracePeriod
  map["isInBillingRetryPeriod"] = this.isInBillingRetryPeriod
  map["isActive"] = this.isActive
  map["expirationDate"] = this.expirationDate?.toIsoString()
  map["store"] = this.store.name
  this.offerType?.let { map["offerType"] = it.toJsonString() }
  this.subscriptionGroupId?.let { map["subscriptionGroupId"] = it }
  return map
}

fun NonSubscriptionTransaction.toJson(): Map<String, Any?> {
  val map = mutableMapOf<String, Any?>()
  map["transactionId"] = this.transactionId
  map["productId"] = this.productId
  map["purchaseDate"] = this.purchaseDate.toIsoString()
  map["isConsumable"] = this.isConsumable
  map["isRevoked"] = this.isRevoked
  map["store"] = this.store.name
  return map
}

File: src/compat/lib/CustomerInfo.ts
import { Entitlement } from "./Entitlement"

/**
 * @category Models
 * Offer type that applied to a {@link SubscriptionTransaction}.
 *
 * iOS only delivers `"trial"`, `"code"`, `"promotional"` and `"winback"`;
 * `"subscription"` and `"revoked"` are Android-only values.
 */
export type SubscriptionOfferType =
  | "trial"
  | "code"
  | "subscription"
  | "promotional"
  | "winback"
  | "revoked"

/**
 * @category Models
 * A subscription transaction in the customer's purchase history.
 */
export interface SubscriptionTransaction {
  transactionId: string
  productId: string
  /** ISO-8601 string. */
  purchaseDate: string
  willRenew: boolean
  isRevoked: boolean
  isInGracePeriod: boolean
  isInBillingRetryPeriod: boolean
  isActive: boolean
  /** ISO-8601 string, or `null` if non-renewing. */
  expirationDate?: string | null
  store: string
  offerType?: SubscriptionOfferType
  subscriptionGroupId?: string
}

/**
 * @category Models
 * A non-subscription (one-time / consumable) transaction.
 */
export interface NonSubscriptionTransaction {
  transactionId: string
  productId: string
  /** ISO-8601 string. */
  purchaseDate: string
  isConsumable: boolean
  isRevoked: boolean
  store: string
}

/**
 * @category Models
 * Snapshot of the customer's subscription and entitlement state.
 * Snapshots are immutable and don't auto-update — fetch a fresh one with
 * `Superwall.shared.getCustomerInfo()` or implement the
 * `customerInfoDidChange` delegate method to stay current.
 */
export class CustomerInfo {
  userId: string
  subscriptions: SubscriptionTransaction[]
  nonSubscriptions: NonSubscriptionTransaction[]
  entitlements: Entitlement[]

  constructor({
    userId,
    subscriptions,
    nonSubscriptions,
    entitlements,
  }: {
    userId: string
    subscriptions: SubscriptionTransaction[]
    nonSubscriptions: NonSubscriptionTransaction[]
    entitlements: Entitlement[]
  }) {
    this.userId = userId
    this.subscriptions = subscriptions
    this.nonSubscriptions = nonSubscriptions
    this.entitlements = entitlements
  }

  static fromJson(json: any): CustomerInfo {
    return new CustomerInfo({
      userId: json.userId ?? "",
      subscriptions: Array.isArray(json.subscriptions) ? json.subscriptions : [],
      nonSubscriptions: Array.isArray(json.nonSubscriptions) ? json.nonSubscriptions : [],
      entitlements: Array.isArray(json.entitlements)
        ? json.entitlements.map((e: any) => Entitlement.fromJson(e))
        : [],
    })
  }
}
